finality-provider/service/fp_instance.go

```diff
@@ -724,12 +733,60 @@ func (fp *FinalityProviderInstance) CommitPubRand(targetBlockHeight uint64) (*ty
 
 	// Update metrics
 	fp.metrics.RecordFpRandomnessTime(fp.GetBtcPkHex())
-	fp.metrics.RecordFpLastCommittedRandomnessHeight(fp.GetBtcPkHex(), lastCommittedHeight)
 	fp.metrics.AddToFpTotalCommittedRandomness(fp.GetBtcPkHex(), float64(len(pubRandList)))
 
 	return res, nil
 }
 
+// TestCommitPubRand is exposed for devops/testing purpose to allow manual committing public randomness in cases
+// where FP is stuck due to lack of public randomness (https://github.com/babylonlabs-io/finality-provider/issues/115).
+//
+// Note:
+// - this function is similar to `CommitPubRand` but should not be used in the main pubrand submission loop.
+// - it will always start from the last committed height + 1
+// - if targetBlockHeight is too large, it will commit multiple fp.cfg.NumPubRand pairs in a loop until reaching the targetBlockHeight
+func (fp *FinalityProviderInstance) TestCommitPubRand(targetBlockHeight uint64) error {
+	lastCommittedHeight, err := fp.GetLastCommittedHeight()
+	if err != nil {
+		return err
+	}
+	fp.metrics.RecordFpLastCommittedRandomnessHeight(fp.GetBtcPkHex(), lastCommittedHeight)
+
+	var startHeight uint64
+	if lastCommittedHeight == uint64(0) {
+		// Note: it can also be the case that the finality-provider has committed 1 pubrand before (but in practice, we
+		// will never set cfg.NumPubRand to 1. so we can safely assume it has never committed before)
+		startHeight = 0
+	} else if lastCommittedHeight < targetBlockHeight {
+		startHeight = lastCommittedHeight + 1
+	} else {
+		return fmt.Errorf(
+			"finality provider has already committed pubrand to target block height (pk: %s, target: %d, last committed: %d)",
+			fp.GetBtcPkHex(),
+			targetBlockHeight,
+			lastCommittedHeight,
+		)
+	}
+	fmt.Printf("Start committing pubrand from block height %d...\n", startHeight)
+
+	commitRandTicker := time.NewTicker(fp.cfg.RandomnessCommitInterval)
+	defer commitRandTicker.Stop()
```

Comment:
I was thinking we should not let user to make the commit with number of randomness larger than the value in config. This way, we can either reject large `target-height` to avoid draining user's funds

Reply:
this command won't be used if it can only commit e.g. 1000 PRs at once. I think the risk is manageable given there is already unsafe prefix and documentation to explain the risk. the idea is anyone who use the command should know exactly what they are doing

Reply:
Agreed but why do we need to set interval? Can the commits be bundled in a single tx with several commit messages?

Reply:
there is already `NumPubRand` in the config. each commit tx will commit this many PRs. the user can change the size in the fpd.conf file.

Reply:
For example, the user will commit `10k` randomness while the config `NumPubRand=1000`, so the cmd will send 10 tx and the user needs to wait for at least `10 * fp.cfg.RandomnessCommitInterval` before it finishes. Why do we want the user to wait for `fp.cfg.RnadomnessCommitInterval` for each commit?

Reply:
makes sense. in this case, the user can set the interval to 0 or sth very small. if I am gonna run it myself, I will probably set it to at least 5 seconds just to be safe

Reply:
maybe it's not necessary but i was thinking about some potential nonce mismatch issue. if we are talking about 2sec block time on the consumer chain, there will be 3600*24/2=43200 blocks per day.

if I set `NumPubRand` to 10000, it should only take 20sec even I add this wait time. and committing PRs for one month's usage would only take 10min. so I guess it's tolerable anyway

Reply:
why would there be nonce mismatch issue? I thought this issue will occur only when we send txs concurrently

Reply:
iiuc, FP interacts with the keyring backend to send txs, if FP sends txs too fast, I am not sure whether there will be any race conditions. maybe the Cosmos keyring infra is quite robust but I am not too familiar w it

Reply:
FP relies on Babylon client implementation which has access control over the keyring `https://github.com/babylonlabs-io/babylon/blob/89dd901bb40dc2420ad196591ec6babe6339cd3c/client/client/tx.go#L50`, so I think it's safe to remove the sending interval. Maybe a better way is to bundle all the commit messages into one like we do for batch finality signatures but I guess we can do it for future optimization

Reply:
got it. I will add a TODO

Reply:
i tested and it doesn't work after removing the interval. i do meet with the nonce mismatch err

```
2024-11-21T01:26:54.093326Z	info	Start committing pubrand from block height	{"start_height": 900}
2024-11-21T01:27:02.632290Z	debug	retrying	{"attemp": 1, "max_attempts": 5, "error": "rpc error: code = Unknown desc = rpc error: code = Unknown desc = account sequence mismatch, expected 11, got 10: incorrect account sequence [cosmos/cosmos-sdk@v0.50.9/x/auth/ante/sigverify.go:290] with gas used: '19626': unknown request"}
2024-11-21T01:27:10.649841Z	debug	retrying	{"attemp": 2, "max_attempts": 5, "error": "rpc error: code = Unknown desc = rpc error: code = Unknown desc = failed to execute message; message index: 0: The start height (900) has overlap with the height of the highest public randomness committed (999): execute wasm contract failed [CosmWasm/wasmd@v0.51.0/x/wasm/keeper/keeper.go:422] with gas used: '114308': unknown request"}
2024-11-21T01:27:19.133247Z	debug	retrying	{"attemp": 3, "max_attempts": 5, "error": "rpc error: code = Unknown desc = rpc error: code = Unknown desc = failed to execute message; message index: 0: The start height (900) has overlap with the height of the highest public randomness committed (999): execute wasm contract failed [CosmWasm/wasmd@v0.51.0/x/wasm/keeper/keeper.go:422] with gas used: '114308': unknown request"}
```

Reply:
It seems the error is because the fpd is already running and committing randomness while you executed the command. This will cause concurrent submission

Reply:
> It seems the error is because the fpd is already running and committing randomness while you executed the command. This will cause concurrent submission

I don't think so. this new command doesn't require `fpd daemon` running

Reply:
I see. But why the account mismatch error occurs at the first submission?

Reply:
I don't know exactly why but i think this is what happened:

- started PR commit at height 900
- first tx is sent for PR 900-999
- before it's processed by the Babylon chain, the second tx for PR 1000-1099 is sent
- but since the first tx is not finished, it reuses the same nonce so we see this error before we see the "Successful transaction" log

Reply:
I see. This is indeed weried as the txs should be sent sequentially. Anyway, let's leave the interval then. Maybe we can also leave a note or TODO there?

Reply:
btw the unit tests failure might be fixed by https://github.com/babylonlabs-io/finality-provider/pull/142

Reply:
approved. i will test again after #142 is merged

Reply:
Thanks. Merged

Reply:
tested. nonce issue is fixed now. thanks!